File: src/crossalpha/observatory/canonical/aave.py
from __future__ import annotations

import gzip
import json
from pathlib import Path
from typing import Any

import pandas as pd

from crossalpha.domain.models import RawSnapshotManifest
from crossalpha.observatory.health import load_manifest
from crossalpha.storage.indexes import load_recent_daily_manifests


LIQUIDATION_COLUMNS = (
    "event_time",
    "observed_at",
    "known_at",
    "block_number",
    "transaction_hash",
    "transaction_index",
    "log_index",
    "block_hash",
    "removed",
    "collateral_asset",
    "debt_asset",
    "user",
    "debt_to_cover_raw",
    "liquidated_collateral_amount_raw",
    "liquidator",
    "receive_atoken",
    "raw_sha256",
    "raw_path",
)


def _to_float(value: Any) -> float | None:
    if value is None or value == "":
        return None
    if isinstance(value, str):
        value = value.strip().replace("%", "").replace(",", "")
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if pd.notna(number) else None


def _to_int_hex(value: Any) -> int | None:
    if not isinstance(value, str):
        return None
    try:
        return int(value, 16)
    except ValueError:
        return None


def _topic_address(value: Any) -> str | None:
    if not isinstance(value, str) or not value.startswith("0x") or len(value) < 42:
        return None
    return "0x" + value[-40:].lower()


def parse_aave_markets(
    envelope: dict[str, Any], raw_record: RawSnapshotManifest
) -> pd.DataFrame:
    payload = envelope.get("payload")
    if not isinstance(payload, dict):
        raise ValueError("Aave markets payload must be an object")
    markets = payload.get("data", {}).get("markets")
    if not isinstance(markets, list) or not markets:
        raise ValueError("Aave markets payload has no markets")

    observed_at = envelope.get("observed_at")
    known_at = envelope.get("known_at")
    metadata = envelope.get("metadata") if isinstance(envelope.get("metadata"), dict) else {}
    chain_id = metadata.get("chain_id")
    rows: list[dict[str, Any]] = []
    for market in markets:
        if not isinstance(market, dict):
            continue
        market_address = market.get("address")
        market_name = market.get("name")
        reserves = market.get("reserves")
        if not isinstance(reserves, list):
            continue
        for reserve in reserves:
            if not isinstance(reserve, dict):
                continue
            token = reserve.get("underlyingToken") or {}
            supply_info = reserve.get("supplyInfo") or {}
            borrow_info = reserve.get("borrowInfo") or {}
            supply_apy = supply_info.get("apy") or {}
            borrow_apy = borrow_info.get("apy") or {}
            available = borrow_info.get("availableLiquidity") or {}
            amount = available.get("amount") or {}
            rows.append(
                {
                    "observed_at": observed_at,
                    "known_at": known_at,
                    "chain_id": chain_id,
                    "market_address": str(market_address).lower() if market_address else None,
                    "market_name": market_name,
                    "reserve_address": str(token.get("address")).lower() if token.get("address") else None,
                    "symbol": token.get("symbol"),
                    "decimals": int(token["decimals"]) if str(token.get("decimals", "")).isdigit() else None,
                    "supply_apy_pct": _to_float(supply_apy.get("formatted")),
                    "borrow_apy_pct": _to_float(borrow_apy.get("formatted")),
                    "available_liquidity_native": _to_float(amount.get("value")),
                    "available_liquidity_usd": _to_float(available.get("usd")),
                    "borrow_cap_reached": bool(borrow_info.get("borrowCapReached", False)),
                    "is_frozen": bool(reserve.get("isFrozen", False)),
                    "is_paused": bool(reserve.get("isPaused", False)),
                    "raw_sha256": raw_record.sha256,
                    "raw_path": raw_record.path,
                }
            )
    frame = pd.DataFrame(rows)
    if frame.empty:
        raise ValueError("Aave markets canonicalization produced zero reserves")
    if frame["symbol"].isna().all():
        raise ValueError("Aave markets canonicalization produced no reserve symbols")
    return frame


def parse_aave_liquidations(
    envelope: dict[str, Any], raw_record: RawSnapshotManifest
) -> pd.DataFrame:
    payload = envelope.get("payload")
    if not isinstance(payload, list):
        raise ValueError("Aave liquidation payload must be a list")
    observed_at = envelope.get("observed_at")
    known_at = envelope.get("known_at")
    rows: list[dict[str, Any]] = []
    for item in payload:
        if not isinstance(item, dict):
            continue
        topics = item.get("topics")
        if not isinstance(topics, list) or len(topics) < 4:
            continue
        data = item.get("data")
        words: list[str] = []
        if isinstance(data, str) and data.startswith("0x"):
            raw = data[2:]
            words = [
                raw[i : i + 64]
                for i in range(0, len(raw), 64)
                if len(raw[i : i + 64]) == 64
            ]
        debt_to_cover_raw = int(words[0], 16) if len(words) > 0 else None
        collateral_amount_raw = int(words[1], 16) if len(words) > 1 else None
        liquidator = "0x" + words[2][-40:].lower() if len(words) > 2 else None
        receive_atoken = bool(int(words[3], 16)) if len(words) > 3 else None
        removed = bool(item.get("removed", False))
        rows.append(
            {
                # Preserve the removed record itself, but remove it from event-time
                # counts after point-in-time deduplication.
                "event_time": None if removed else item.get("blockTimestamp"),
                "observed_at": observed_at,
                "known_at": known_at,
                "block_number": _to_int_hex(item.get("blockNumber")),
                "transaction_hash": item.get("transactionHash"),
                "transaction_index": _to_int_hex(item.get("transactionIndex")),
                "log_index": _to_int_hex(item.get("logIndex")),
                "block_hash": item.get("blockHash"),
                "removed": removed,
                "collateral_asset": _topic_address(topics[1]),
                "debt_asset": _topic_address(topics[2]),
                "user": _topic_address(topics[3]),
                "debt_to_cover_raw": debt_to_cover_raw,
                "liquidated_collateral_amount_raw": collateral_amount_raw,
                "liquidator": liquidator,
                "receive_atoken": receive_atoken,
                "raw_sha256": raw_record.sha256,
                "raw_path": raw_record.path,
            }
        )
    # An empty liquidation scan is valid evidence (zero observed events), not an
    # absent schema. Keeping fixed columns lets parquet/catalog readers distinguish
    # a successful zero-event scan from a missing collector.
    return pd.DataFrame(rows, columns=LIQUIDATION_COLUMNS)


def _records(data_root: Path, recent_days: int | None) -> tuple[list[RawSnapshotManifest], str]:
    if recent_days is None:
        records, errors = load_manifest(data_root)
        mode = "full"
    else:
        records, errors = load_recent_daily_manifests(data_root, days=recent_days)
        mode = f"recent_{recent_days}d"
    if errors:
        raise ValueError(f"raw manifest contains errors: {errors}")
    return records, mode


def canonicalize_aave(
    data_root: Path,
    *,
    recent_days: int | None = None,
) -> dict[str, int | str]:
    records, mode = _records(data_root, recent_days)
    market_records = [
        row
        for row in records
        if row.source_id == "aave:v3:graphql" and row.observation_type == "markets_snapshot"
    ]
    liquidation_records = [
        row
        for row in records
        if row.source_id == "aave:v3:ethereum" and row.observation_type == "liquidation_logs"
    ]

    market_written = market_skipped = market_rows = 0
    liquidation_written = liquidation_skipped = liquidation_rows = 0

    for record in sorted(market_records, key=lambda item: item.observed_at):
        observed = record.observed_at
        out_dir = (
            data_root
            / "canonical"
            / "aave"
            / "markets"
            / f"year={observed:%Y}"
            / f"month={observed:%m}"
            / f"day={observed:%d}"
        )
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{observed:%Y%m%dT%H%M%S.%fZ}_{record.sha256[:12]}.parquet"
        if out_path.exists():
            market_skipped += 1
            continue
        with gzip.open(record.path, "rt", encoding="utf-8") as fh:
            envelope = json.load(fh)
        frame = parse_aave_markets(envelope, record)
        tmp = out_path.with_suffix(".parquet.tmp")
        frame.to_parquet(tmp, index=False)
        tmp.replace(out_path)
        market_written += 1
        market_rows += len(frame)

    for record in sorted(liquidation_records, key=lambda item: item.observed_at):
        observed = record.observed_at
        out_dir = (
            data_root
            / "canonical"
            / "aave"
            / "liquidations"
            / f"year={observed:%Y}"
            / f"month={observed:%m}"
            / f"day={observed:%d}"
        )
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{observed:%Y%m%dT%H%M%S.%fZ}_{record.sha256[:12]}.parquet"
        if out_path.exists():
            liquidation_skipped += 1
            continue
        with gzip.open(record.path, "rt", encoding="utf-8") as fh:
            envelope = json.load(fh)
        frame = parse_aave_liquidations(envelope, record)
        tmp = out_path.with_suffix(".parquet.tmp")
        frame.to_parquet(tmp, index=False)
        tmp.replace(out_path)
        liquidation_written += 1
        liquidation_rows += len(frame)

    return {
        "mode": mode,
        "market_snapshots": len(market_records),
        "market_written": market_written,
        "market_skipped": market_skipped,
        "market_rows_written": market_rows,
        "liquidation_snapshots": len(liquidation_records),
        "liquidation_written": liquidation_written,
        "liquidation_skipped": liquidation_skipped,
        "liquidation_rows_written": liquidation_rows,
    }

File: crates/crossalpha-features/src/canonical/aave.rs
use anyhow::{Context, Result, bail};
use chrono::{DateTime, Utc};
use crossalpha_storage::{ObservationEnvelope, RawSnapshotManifest};
use serde::Serialize;
use serde_json::Value;

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AaveMarketRow {
    pub observed_at: DateTime<Utc>,
    pub known_at: DateTime<Utc>,
    pub chain_id: Option<i64>,
    pub market_address: Option<String>,
    pub market_name: Option<String>,
    pub reserve_address: Option<String>,
    pub symbol: Option<String>,
    pub decimals: Option<i64>,
    pub supply_apy_pct: Option<f64>,
    pub borrow_apy_pct: Option<f64>,
    pub available_liquidity_native: Option<f64>,
    pub available_liquidity_usd: Option<f64>,
    pub borrow_cap_reached: bool,
    pub is_frozen: bool,
    pub is_paused: bool,
    pub raw_sha256: String,
    pub raw_path: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AaveLiquidationRow {
    pub event_time: Option<DateTime<Utc>>,
    pub observed_at: DateTime<Utc>,
    pub known_at: DateTime<Utc>,
    pub block_number: Option<u64>,
    pub transaction_hash: Option<String>,
    pub transaction_index: Option<u64>,
    pub log_index: Option<u64>,
    pub block_hash: Option<String>,
    pub removed: bool,
    pub collateral_asset: Option<String>,
    pub debt_asset: Option<String>,
    pub user: Option<String>,
    /// Decimal uint256 text. State V0.2 never treats raw token units as floating point.
    pub debt_to_cover_raw: Option<String>,
    /// Decimal uint256 text. State V0.2 never treats raw token units as floating point.
    pub liquidated_collateral_amount_raw: Option<String>,
    pub liquidator: Option<String>,
    pub receive_atoken: Option<bool>,
    pub raw_sha256: String,
    pub raw_path: String,
}

pub fn parse_aave_markets(
    envelope: &ObservationEnvelope,
    raw_record: &RawSnapshotManifest,
) -> Result<Vec<AaveMarketRow>> {
    let payload = envelope
        .payload
        .as_object()
        .context("Aave markets payload must be an object")?;
    let markets = payload
        .get("data")
        .and_then(Value::as_object)
        .and_then(|data| data.get("markets"))
        .and_then(Value::as_array)
        .context("Aave markets payload has no markets")?;
    if markets.is_empty() {
        bail!("Aave markets payload has no markets");
    }

    let chain_id = envelope.metadata.get("chain_id").and_then(value_i64);
    let mut rows = Vec::new();
    for market in markets {
        let Some(market) = market.as_object() else {
            continue;
        };
        let reserves = market
            .get("reserves")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        for reserve in reserves {
            let Some(reserve) = reserve.as_object() else {
                continue;
            };
            let token = reserve.get("underlyingToken").and_then(Value::as_object);
            let supply = reserve.get("supplyInfo").and_then(Value::as_object);
            let borrow = reserve.get("borrowInfo").and_then(Value::as_object);
            let supply_apy = supply
                .and_then(|value| value.get("apy"))
                .and_then(Value::as_object)
                .and_then(|value| value.get("formatted"))
                .and_then(to_float);
            let borrow_apy = borrow
                .and_then(|value| value.get("apy"))
                .and_then(Value::as_object)
                .and_then(|value| value.get("formatted"))
                .and_then(to_float);
            let available = borrow
                .and_then(|value| value.get("availableLiquidity"))
                .and_then(Value::as_object);
            let native = available
                .and_then(|value| value.get("amount"))
                .and_then(Value::as_object)
                .and_then(|value| value.get("value"))
                .and_then(to_float);
            let usd = available
                .and_then(|value| value.get("usd"))
                .and_then(to_float);

            rows.push(AaveMarketRow {
                observed_at: envelope.observed_at,
                known_at: envelope.known_at,
                chain_id,
                market_address: market
                    .get("address")
                    .and_then(Value::as_str)
                    .map(|value| value.to_ascii_lowercase()),
                market_name: market
                    .get("name")
                    .and_then(Value::as_str)
                    .map(str::to_owned),
                reserve_address: token
                    .and_then(|value| value.get("address"))
                    .and_then(Value::as_str)
                    .map(|value| value.to_ascii_lowercase()),
                symbol: token
                    .and_then(|value| value.get("symbol"))
                    .and_then(Value::as_str)
                    .map(str::to_owned),
                decimals: token
                    .and_then(|value| value.get("decimals"))
                    .and_then(value_i64),
                supply_apy_pct: supply_apy,
                borrow_apy_pct: borrow_apy,
                available_liquidity_native: native,
                available_liquidity_usd: usd,
                borrow_cap_reached: borrow
                    .and_then(|value| value.get("borrowCapReached"))
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
                is_frozen: reserve
                    .get("isFrozen")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
                is_paused: reserve
                    .get("isPaused")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
                raw_sha256: raw_record.sha256.clone(),
                raw_path: raw_record.path.clone(),
            });
        }
    }
    if rows.is_empty() {
        bail!("Aave markets canonicalization produced zero reserves");
    }
    if rows.iter().all(|row| row.symbol.is_none()) {
        bail!("Aave markets canonicalization produced no reserve symbols");
    }
    Ok(rows)
}

pub fn parse_aave_liquidations(
    envelope: &ObservationEnvelope,
    raw_record: &RawSnapshotManifest,
) -> Result<Vec<AaveLiquidationRow>> {
    let payload = envelope
        .payload
        .as_array()
        .context("Aave liquidation payload must be a list")?;
    let mut rows = Vec::new();
    for item in payload {
        let Some(item) = item.as_object() else {
            continue;
        };
        let topics = item
            .get("topics")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        if topics.len() < 4 {
            continue;
        }
        let words = data_words(item.get("data"));
        let removed = item
            .get("removed")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        let event_time = if removed {
            None
        } else {
            item.get("blockTimestamp")
                .and_then(Value::as_str)
                .and_then(parse_datetime)
        };
        rows.push(AaveLiquidationRow {
            event_time,
            observed_at: envelope.observed_at,
            known_at: envelope.known_at,
            block_number: item.get("blockNumber").and_then(hex_u64),
            transaction_hash: item
                .get("transactionHash")
                .and_then(Value::as_str)
                .map(str::to_owned),
            transaction_index: item.get("transactionIndex").and_then(hex_u64),
            log_index: item.get("logIndex").and_then(hex_u64),
            block_hash: item
                .get("blockHash")
                .and_then(Value::as_str)
                .map(str::to_owned),
            removed,
            collateral_asset: topic_address(topics.get(1)),
            debt_asset: topic_address(topics.get(2)),
            user: topic_address(topics.get(3)),
            debt_to_cover_raw: words.first().and_then(|word| hex_uint_decimal(word)),
            liquidated_collateral_amount_raw: words.get(1).and_then(|word| hex_uint_decimal(word)),
            liquidator: words.get(2).map(|word| {
                format!("0x{}", &word[word.len().saturating_sub(40)..]).to_ascii_lowercase()
            }),
            receive_atoken: words.get(3).and_then(|word| {
                u8::from_str_radix(&word[word.len().saturating_sub(2)..], 16)
                    .ok()
                    .map(|value| value != 0)
            }),
            raw_sha256: raw_record.sha256.clone(),
            raw_path: raw_record.path.clone(),
        });
    }
    Ok(rows)
}

fn to_float(value: &Value) -> Option<f64> {
    let number = if let Some(number) = value.as_f64() {
        number
    } else if let Some(text) = value.as_str() {
        text.trim()
            .replace('%', "")
            .replace(',', "")
            .parse::<f64>()
            .ok()?
    } else {
        return None;
    };
    number.is_finite().then_some(number)
}

fn value_i64(value: &Value) -> Option<i64> {
    value
        .as_i64()
        .or_else(|| value.as_u64().and_then(|value| i64::try_from(value).ok()))
        .or_else(|| value.as_str().and_then(|value| value.parse::<i64>().ok()))
}

fn hex_u64(value: &Value) -> Option<u64> {
    let text = value.as_str()?.strip_prefix("0x")?;
    u64::from_str_radix(text, 16).ok()
}

fn topic_address(value: Option<&Value>) -> Option<String> {
    let text = value?.as_str()?;
    let raw = text.strip_prefix("0x")?;
    if raw.len() < 40 || !raw.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return None;
    }
    Some(format!("0x{}", &raw[raw.len() - 40..]).to_ascii_lowercase())
}

fn data_words(value: Option<&Value>) -> Vec<String> {
    let Some(raw) = value
        .and_then(Value::as_str)
        .and_then(|value| value.strip_prefix("0x"))
    else {
        return Vec::new();
    };
    raw.as_bytes()
        .chunks(64)
        .filter(|chunk| chunk.len() == 64)
        .filter_map(|chunk| std::str::from_utf8(chunk).ok().map(str::to_owned))
        .collect()
}

fn parse_datetime(value: &str) -> Option<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(value)
        .ok()
        .map(|value| value.with_timezone(&Utc))
}

fn hex_uint_decimal(word: &str) -> Option<String> {
    let value = num_bigint::BigUint::parse_bytes(word.as_bytes(), 16)?;
    Some(value.to_str_radix(10))
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;
    use crossalpha_storage::ObservationEnvelope;
    use serde_json::json;

    fn record() -> RawSnapshotManifest {
        RawSnapshotManifest {
            path: "/tmp/aave.json.gz".to_owned(),
            sha256: "abc".to_owned(),
            bytes: 10,
            compressed_bytes: Some(5),
            observed_at: Utc.timestamp_opt(1_700_000_000, 0).unwrap(),
            source_id: "aave:v3:graphql".to_owned(),
            observation_type: "markets_snapshot".to_owned(),
        }
    }

    #[test]
    fn parses_market_reserve_fields_without_schema_drift() {
        let at = Utc.timestamp_opt(1_700_000_000, 0).unwrap();
        let envelope = ObservationEnvelope {
            schema_version: 1,
            event_time: None,
            observed_at: at,
            known_at: at,
            source_type: "AGGREGATOR".to_owned(),
            source_id: "aave:v3:graphql".to_owned(),
            observation_type: "markets_snapshot".to_owned(),
            payload: json!({"data":{"markets":[{"address":"0xABC","name":"Core","reserves":[{"underlyingToken":{"address":"0xDEF","symbol":"WETH","decimals":18},"supplyInfo":{"apy":{"formatted":"2.0"}},"borrowInfo":{"apy":{"formatted":"5.0"},"availableLiquidity":{"amount":{"value":"12"},"usd":"100"},"borrowCapReached":false},"isFrozen":false,"isPaused":false}]}]}}),
            metadata: serde_json::Map::from_iter([("chain_id".to_owned(), json!(1))]),
        };
        let rows = parse_aave_markets(&envelope, &record()).unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0].symbol.as_deref(), Some("WETH"));
        assert_eq!(rows[0].available_liquidity_usd, Some(100.0));
    }
}
